### DISCUSSION TITLE
gamestate/state-machine with help of variable holding function names

### DISCUSSION THREAD

**Discussion by nikgid** [nesbox/TIC-80#2075] [Q&A] [answered]
I want to implement a simple gamestate-system, similar to the one described here: https://pico-8.fandom.com/wiki/Yet_Another_Multiple_States_System – it uses the "trick" to let a variable hold function names and then call the variable as a function.
This should be possible in Lua (also according to: https://www.lua.org/pil/6.html)
But my simple example here:

```lua
gamestate=title

function TIC()
        gamestate()
end

function title()
	cls(2)
	print("title gamestate",10,10,12)
	if btnp(4) then
		init_play()
	end
end

function init_play()
	gamestate=play
end

function play()
	cls(6)
	print("play gamestate",10,10,12)
	if btnp(4) then
		init_title()
	end
end

function init_title()
	gamestate=title
end
```

... does not work but throws an error
`[string "gamestate=title..."]:4: 
attempt to call a nil value (global 
'gamestate')
stack traceback:
	[string "gamestate=title..."]:4: in 
function 'TIC'`

What am I doing wrong here?

**Comment by nikgid** [comment]
Okay, I tested a bit more and found this somehow "dirty" solution – order of function definitions seems to matter. This version, where I define the different gamestate-functions _before_ applying one of them to the variable "gamestate", works!

```lua
function title()
	cls(2)
	print("title gamestate",10,10,12)
	if btnp(4) then
		init_play()
	end
end

function play()
	cls(6)
	print("play gamestate",10,10,12)
	if btnp(4) then
		init_title()
	end
end

gamestate=title

function TIC()
	gamestate()
end

function init_play()
	gamestate=play
end

function init_title()
	gamestate=title
end
```

But still this seems to be a bit awkward ... isn't there a better way to write the line ```gamestate = title``` – in another way so that it is clear that "gamestate" is a placeholder for function-names, starting with the function"title" (which is to be defined later)?

**Reply by Anrock** [reply]
It's not dirty, it's actually how some interpreters (and some non-interpreted languages) work - you can't use a thing before you declare a thing. That's why it's common and good practice to put declarations on top of the file and executable line after them.

**Reply by nikgid** [reply]
Thank you for your answer! I was looking for a solution without a need for a strict order of the function definitions, because I am using TIC-80 (and Pico-8) for coding with younger kids. And being free from as much "abstract" concepts of order and/or syntax as possible is one of the reasons I prefer Lua/TIC-80/Pico-8 for this purpose. It leaves room/time to focus on the more concrete/graspable coding concepts first.

**Comment by joshgoebel** [comment]
Use the BOOT function for startup code. https://github.com/nesbox/TIC-80/wiki/BOOT

**Reply by nikgid** [reply]
Thanks a lot! (I saw this after posting my solution.)

**Comment by nikgid** [comment, marked as answer]
I found out a solution that seems to be the best / most logical for me:
The built-in BOOT-function seems to be called as the first function but **after** TIC-80 has "had a look at" the rest of the code. 
Thus putting `gamestate=title` inside BOOT also works and I can happily put the function "title" wherever I want.
My solution:

```lua
function BOOT()
	gamestate=title
end

function TIC()
	gamestate()
end

function init_play()
	gamestate=play
end


function init_title()
	gamestate=title
end

function title()
	cls(2)
	print("title gamestate",10,10,12)
	if btnp(4) then
		init_play()
	end
end

function play()
	cls(6)
	print("play gamestate",10,10,12)
	if btnp(4) then
		init_title()
	end
end
```
